libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,955 @@
+		{"textDocument/documentHighlight", bind(&LanguageServer::handle_textDocument_highlight, this, _1, _2)},
+		{"textDocument/references", bind(&LanguageServer::handle_textDocument_references, this, _1, _2)},
+	},
+	m_logger{std::move(_logger)}
+{
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
+{
+	if (_settings["evm"].isString())
+		if (auto const evmVersionOpt = EVMVersion::fromString(_settings["evm"].asString()); evmVersionOpt.has_value())
+			m_evmVersion = evmVersionOpt.value();
+
+	if (_settings["remapping"].isArray())
+	{
+		for (auto const& element: _settings["remapping"])
+		{
+			if (element.isString())
+			{
+				if (auto remappingOpt = CompilerStack::parseRemapping(element.asString()); remappingOpt.has_value())
+					m_remappings.emplace_back(move(remappingOpt.value()));
+				else
+					trace("Failed to parse remapping: '"s + element.asString() + "'");
+			}
+		}
+	}
+}
+
+namespace { // TODO: move us up!
+	inline size_t toOffset(std::string const& _text, langutil::LineColumn _position) noexcept
+	{
+		// TODO: take care of Unicode.
+		size_t offset = 0;
+		langutil::LineColumn current = {0, 0};
+		while (current != _position && offset < _text.size())
+		{
+			if (_text.at(offset) != '\n')
+				current.column++;
+			else
+			{
+				current.line++;
+				current.column = 0;
+			}
+			offset++;
+		}
+		return offset;
+	}
+
+	inline std::pair<size_t, size_t> offsetsOf(std::string const& _text, LineColumnRange _range) noexcept
+	{
+		return std::pair{
+			toOffset(_text, _range.start),
+			toOffset(_text, _range.end)
+		};
+	}
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, LineColumnRange _range, string const& _replacementText)
+{
+	// TODO: all this info is actually unrelated to solidity/lsp specifically except knowing that
+	// the file has updated, so we can  abstract that away and only do the re-validation here.
+	auto file = m_fileReader->sourceCodes().find(_path);
+	if (file == m_fileReader->sourceCodes().end())
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	auto& buffer = file->second;
+
+	auto const [start, end] = offsetsOf(buffer, _range);
+	buffer.replace(start, end - start, _replacementText);
+	buffer.replace(start, end - start, _replacementText);
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, string const& _replacementText)
+{
+	auto file = m_fileReader->sourceCodes().find(_path);
+	if (file == m_fileReader->sourceCodes().end())
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	file->second = _replacementText;
+
+	validate(_path);
+}
+
+frontend::ReadCallback::Result LanguageServer::readFile(string const& _kind, string const& _path)
+{
+	return m_fileReader->readFile(_kind, _path);
+}
+
+constexpr DiagnosticSeverity toDiagnosticSeverity(Error::Type _errorType)
```

Comment:
I think this can be seriously simplifying removing the need for the enum. If you want, you can define the enum locally where this is used (in the conversion to json).